implement pushactivetobottom dispacher

alaricljs opened the pull request:
Adds pushactivetobottom dispacher to manage Z-order of floating windows, opposite of bringactivetotop.
Makes both bringactivetotop and pushactivetobottom available to mouse binds.

I had issues with Compositor.cpp:1255 - making this generic with less code via:

```
auto begin = top ? m_vWindows.begin() : (m_vWindows.rbegin() + 1).base();
auto end = top ? m_vWindows.end() : m_vWindows.rend();
```

Never worked for `top=false` regardless of variations to m_vWindows.rbegin().  Being new to C++ I don't know if it's  because of the uses of begin & end inside the algo, or just my botching the cast of the reverse iterator.


In KeybindManager.cpp:1983 I kludged in mouse re-focus behavior without checking follow_mouse or mouse_refocus.  Is this the right place for it?  It doesn't refocus without the simulateMouseMovement() call, but with that call it doesn't obey those 2 options which I would have thought handled in InputManager.cpp around ln 400.

Need to resolve the above with your input, thanks.

vaxerski reviewed (commented):
hm. Maybe we could unify both into one dispatcher?

Like, "alterzorder" with either "+x" "-x" or "top" "bottom"

We'd deprecate bringactivetotop by that, too.

alaricljs commented:
I am a fan of good unified design and deprecating the outdated.  One thing though, I can't think of a reason for +/- N changes, even when it was available I only ever used Top/Bottom.  No issue with implementing it if you want.

vaxerski commented:
maybe we can put it off then for now, and if someone requests it and it's desired, then add it.

alaricljs commented:
Implemented alterzorder which allows for a window filter but defaults to the active window.  Also removed my adds to the mouse bind.  I managed to miss that short blurb that bind can handle mouse buttons for clicks.

vaxerski commented on src/managers/KeybindManager.cpp:
```diff
@@ -1980,12 +1972,29 @@ void CKeybindManager::bringActiveToTop(std::string args) {
+        return;
     }
+
+    // if follow_mouse && mouse_refocus ?
```
nope, this is to unstuck the cursor. comment can be removed

vaxerski commented on src/managers/KeybindManager.cpp:
```diff
@@ -1980,12 +1972,29 @@ void CKeybindManager::bringActiveToTop(std::string args) {
+    const auto POSITION    = args.substr(0, args.find_first_of(','));
+    auto       PWINDOW     = g_pCompositor->getWindowByRegex(WINDOWREGEX);
+
+    if (!PWINDOW && g_pCompositor->m_pLastWindow && g_pCompositor->m_pLastWindow->m_bIsFloating)
```
this will fail if provided with a window, and crash if m_pLastWindow is null.

```cpp
if (!PWINDOW && g_pCompositor->m_pLastWindow && g_pCompositor->m_pLastWindow->m_bIsFloating)
    PWINDOW = g_pCompositor->m_pLastWindow;

if (!PWINDOW) {
    Debug::log(ERR, "alterZOrder: no window");
    return;
}
```

vaxerski commented on src/Compositor.cpp:
```diff
@@ -1278,9 +1292,9 @@ void CCompositor::moveWindowToTop(CWindow* pWindow) {
+        moveToZ(pw, top);
 
-        moveWindowToTop(pw);
+        changeWindowZOrder(pw, top);
```
both of these should be true, not top

alaricljs replied:
Nope, it's uglier than that... 
For the X11 windows I'll need to figure out what order toMove has the windows in and then move them to the bottom Z from the top of the stack down.   I'm thinking the original is just doing a 'top' form the bottom of the stack up.  It'll take me more than the time I have tonight to fix that and test it.

vaxerski replied:
parent -> child -> child


you move the parent to the bottom, but children should still be preserved on top of it.

You're right it shouldn't be `true`, my bad. it should be children from the bottom first to the bottom, parent last, lol

vaxerski commented on src/Compositor.cpp:
```diff
@@ -1245,24 +1245,38 @@ bool CCompositor::isWindowActive(CWindow* pWindow) {
 }
 
 void CCompositor::moveWindowToTop(CWindow* pWindow) {
+    changeWindowZOrder(pWindow, true);
```
why not rid of this? internal func, we can just yeet it and replace the refs with changewindowzorder

alaricljs commented:
It's solved... but it's ugly.  I'll think on it for a bit.

alaricljs commented:
Ok, that's less ugly

alaricljs commented:
https://github.com/hyprwm/hyprland-wiki/pull/337

vaxerski commented:
my bad

MightyPlaza commented:
> I am a fan of good unified design and deprecating the outdated. One thing though, I can't think of a reason for +/- N changes, even when it was available I only ever used Top/Bottom. No issue with implementing it if you want.

when having grouped or other hidden windows it would be best to have to option to swap/set "the same" z order when changing the visible window
otherwise each window will have it's own z order or changing visible window will require moving it to the top
